pkg/reconciler/taskrun/validate_resources_test.go

```diff
@@ -550,3 +556,65 @@ func TestValidateOverrides(t *testing.T) {
 		})
 	}
 }
+
+func TestCorrectAoSValueType(t *testing.T) {
+	tcs := []struct {
+		name          string
+		params        []v1beta1.Param
+		neededTypes   map[string]v1beta1.ParamType
+		expectedTypes map[string]v1beta1.ParamType
+	}{{
+		name: "given string need string",
+		params: []v1beta1.Param{
+			{
+				Name:  "foo",
+				Value: *v1beta1.NewArrayOrString("bar"),
+			},
+		},
+		neededTypes: map[string]v1beta1.ParamType{
+			"foo": v1beta1.ParamTypeString,
+		},
+		expectedTypes: map[string]v1beta1.ParamType{
+			"foo": v1beta1.ParamTypeString,
+		},
+	}, {
+		name: "given string needed array but param value is not reference",
+		params: []v1beta1.Param{
+			{
+				Name:  "foo",
+				Value: *v1beta1.NewArrayOrString("bar"),
+			},
+		},
+		neededTypes: map[string]v1beta1.ParamType{
+			"foo": v1beta1.ParamTypeArray,
+		},
+		expectedTypes: map[string]v1beta1.ParamType{
+			"foo": v1beta1.ParamTypeString,
+		},
+	}, {
+		name: "given string needed array and param value is reference",
+		params: []v1beta1.Param{
+			{
+				Name:  "foo",
+				Value: *v1beta1.NewArrayOrString("$(results.resultname[*])"),
```

Comment:
Also, if the string is `"$(params.stringA) followed by $(params.stringB)"`, `VariableSubstitutionRegex.MatchString` also returns true and the test passes. Is it expected to also **correct** the type in such case?

Reply:
> @Yongxuanzhang if I change `"$(results.resultname[*])"` to `"$(results.resultname)"`, the test still passes. Is that expected? Is there another place that we validate that without a `[*]`, the param should be invalid?

It is supported by the following function so it can pass, for actual use since I didn't add this type, this is not working actually. I'm not sure if we should support `"$(results.resultname)"` for array and object substitution. 
https://github.com/tektoncd/pipeline/blob/3147117bda072f801f03268eb2a6d1379a692962/pkg/substitution/substitution.go#L282-L290

Reply:
tep 0076 mentions that we should use `*` so we shouldn't support this. It looks like a bit controversial here. 
https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md#to-provide-values-for-array-params

Reply:
> Also, if the string is `"$(params.stringA) followed by $(params.stringB)"`, `VariableSubstitutionRegex.MatchString` also returns true and the test passes. Is it expected to also **correct** the type in such case?

discussed offline about this, will leave to @chuangw6's PR to handle type stuff. We will skip some validation cases here

Reply:
Maybe leave this case as an issue and fix this in another pr.

Reply:
yes, it will be our cleanup work

Reply:
Can you confirm that this is being handled, as you already rebased the PR?

Reply:
ok the test here still passes if I replace $(results.foo[*]) with $(results.foo). Could we add some validation for this or at least a comment that this is upcoming work.

Reply:
After #5007 was merged, the current code supports $(results.foo) to refer to an array. I left one todo in this PR to be validate this use case and it should be invalid.